flink-runtime/src/main/java/org/apache/flink/runtime/jobmaster/factories/DefaultJobManagerJobMetricGroupFactory.java

```diff
@@ -0,0 +1,46 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.flink.runtime.jobmaster.factories;
+
+import org.apache.flink.api.common.JobID;
+import org.apache.flink.runtime.jobgraph.JobGraph;
+import org.apache.flink.runtime.metrics.groups.JobManagerJobMetricGroup;
+import org.apache.flink.runtime.metrics.groups.JobManagerMetricGroup;
+
+import javax.annotation.Nonnull;
+
+/**
+ * Implementation of {@link JobManagerJobMetricGroupFactory} which makes sure that there is always
+ * at most one {@link JobManagerJobMetricGroup} for a given {@link JobID} registered.
+ */
+public class DefaultJobManagerJobMetricGroupFactory implements JobManagerJobMetricGroupFactory {
+
+	private final JobManagerMetricGroup jobManagerMetricGroup;
+
+	public DefaultJobManagerJobMetricGroupFactory(@Nonnull JobManagerMetricGroup jobManagerMetricGroup) {
+		this.jobManagerMetricGroup = jobManagerMetricGroup;
+	}
+
+	@Override
+	public JobManagerJobMetricGroup create(@Nonnull JobGraph jobGraph) {
+		jobManagerMetricGroup.removeJob(jobGraph.getJobID());
```

Comment:
This should only be necessary if the group wasn't properly cleaned up previously by the JobMaster, correct? If so I suggest for debugging purposes to modify `removeJob` to return a boolean indicating whether something was actually deleted and log that as a warning.

Reply:
Yes, but the `JobMaster` does not have access to the `JobManagerMetricGroup`. That's why it is necessary to remove it here. Can change the return type of `JobManagerMetricGroup#removeJob` but I think it is not strictly necessary.

Reply:
We could also pass in the `JobManagerMetricGroup` into the `JobMaster`, but I think by only exposing the `JobManagerJobMetricGroup` to it, gives a better separation of concerns.

Reply:
you could use the same pattern we use for `TaskMetricGroups`, which remove themselves from their parent `TaskManagerJobMetricGroup` when they are closed. That said I'm not really fond of it.

You could also add a `reset` method to the `JobManagerJobMetricGroup` that removes all metrics. This may be more reasonable than replacing the group.

Reply:
I think closing the `JobManagerJobMetricGroup` alone is not enough, because then we still have to create a fresh one.

I was a bit hesitant changing the `AbstractMetricGroup` because to me it looked as if you could only add metrics but never remove them (except when closing). So in order to be as little invasive as possible, I went for the approach where we create a fresh metrics group whenever we recreate the `ExecutionGraph`.

Given the factory, this could also be easily changed later because whether you create a fresh instance or reset it is an implementation detail.

Reply:
yes, you would still need the factory when using the `TaskMetricGroup`-like pattern, but then
* the factory wouldn't have that odd side-effect
* there would be a single _expected_ path on which metrics are unregistered (currently, the last `JobManagerMetricGroup` is implicitly closed when the `JobManagerMetricGroup` is closed)

Reply:
You're right, I will change it like you've described:

- Call `JobManagerJobMetricGroup#close` explicitly
- Remove `JobManagerMetricGroup#removeJob` from `DefaultJobManagerJobMetricGroupFactory`